lib/rouge/lexers/solidity.rb

```diff
@@ -0,0 +1,190 @@
+# -*- coding: utf-8 -*- #
+
+module Rouge
+  module Lexers
+    class Solidity < RegexLexer
+      title "Solidity"
+      desc "Solidity, an Ethereum smart contract programming language"
+      tag 'solidity'
+      filenames '*.sol', '*.solidity'
+      mimetypes 'text/x-solidity'
+
+      # optional comment or whitespace
+      ws = %r((?:\s|//.*?\n|/[*].*?[*]/)+)
+      id = /[a-zA-Z$_][a-zA-Z0-9$_]*/
+
+      def self.detect?(text)
+        return true if text.start_with? 'pragma solidity'
+      end
+
+      # TODO: seperate by "type"
+      def self.keywords
+        @keywords ||= Set.new %w(
+          abstract anonymous as assembly break catch constant constructor continue
+          contract do delete else emit enum event external fallback for function hex
+          if indexed interface internal import is library mapping memory
+          modifier new payable public pure pragma private return returns
+          storage struct throw try type using var view while
+        )
+      end
+
+      def self.builtins
+        return @builtins if @builtins
+        @builtins = Set.new %w(
+          now
+          false true
+          balance now selector super this
+          blockhash gasleft
+          assert require revert
+          selfdestruct suicide
+          call callcode delegatecall
+          send transfer
+          addmod ecrecover keccak256 mulmod sha256 sha3 ripemd160
+        )
+        # TODO: use (currently shadowed by catch-all in :statements)
+        abi = %w(encode encodePacked encodeWithSelector encodeWithSignature)
+        @builtins.merge( abi.map { |i| "abi.#{i}" } )
+        block = %w(coinbase difficulty gaslimit hash number timestamp)
+        @builtins.merge( block.map { |i| "block.#{i}" } )
+        msg = %w(data gas sender sig value)
+        @builtins.merge( msg.map { |i| "msg.#{i}" } )
+        tx = %w(gasprice origin)
+        @builtins.merge( tx.map { |i| "tx.#{i}" } )
+      end
+
+      def self.constants
+        @constants ||= Set.new %w(
+          wei finney szabo ether
+          seconds minutes hours days weeks years
+        )
+      end
+
+      def self.keywords_type
+        return @keywords_type if @keywords_type
+        @keywords_type = Set.new %w(
+          address bool bytes fixed int ufixed uint string
+        )
+
+        # bytes1 .. bytes32
+        @keywords_type.merge( (1..32).map { |i| "bytes#{i}" } )
+
+        # size helpers
+        sizesm = (0..256).step(8)
+        sizesn = (8..256).step(8)
+        sizesmxn = sizesm.map { |m| m }
+                     .product( sizesn.map { |n| n } )
+                     .select { |m,n| m+n <= 256 }
+        # [u]int8 .. [u]int256
+        @keywords_type.merge( sizesn.map { |n|  "int#{n}" } )
+        @keywords_type.merge( sizesn.map { |n| "uint#{n}" } )
+        # [u]fixed{MxN}
+        @keywords_type.merge(sizesmxn.map { |m,n|  "fixed#{m}x#{n}" })
+        @keywords_type.merge(sizesmxn.map { |m,n| "ufixed#{m}x#{n}" })
```

Comment:
I think this logic `sizesmxn` is outdated. It is based on binary points, but the latest was based on decimal points. However it is an unsupported feature as of now, so perhaps I would suggest just to avoid `ufixed*`.

Reply:
They're still [listed in the docs](https://solidity.readthedocs.io/en/v0.6.4/abi-spec.html#types) (where I lifted them from), but OK, I see [the tracking issue](https://github.com/ethereum/solidity/issues/409) is still open... From skimming it, and searching the repo for `FixedPoint`, it seems that `fixed` is also not entirely implemented, right? Should that, too, be removed?

Reply:
The compiler only supports declaration of the type, but nothing else, so it is practically useless. And if you check tracking issue, even basic decisions are still not fully covered.

It doesn't matter too much, we plan to make a decision in the coming month on it, but I would consider it as perhaps like reserved keywords and nothing more.

Reply:
What's the best way to resolve this?

Reply:
Almost done with a commit that moves the `fixed`/`ufixed` to the `reserved` keywords section.

The current implementation in the lexer follows what's in the docs, so if someone was to try and use it, that's what they'd probably try.

I think that's good enough for now.

Reply:
I've tried "fixing" it to the current scheme supported by the compiler, where the MxN types would need to have

``` ruby
        sizesm = (8..256).step(8)
        sizesn = (0..80)
```

... and that would produce 5022 new reserved keywords to be added for the `[u]fixed{M}x{N}` types.

Is this acceptable?

The `diff` includes one more line, as shown [here](https://gist.github.com/veox/7810938e8559994f207a036b2362ac72): removing the `.product()` line.

Reply:
Hmm, the current scheme produced 1984 type keywords, so yeah, I guess that'll do. It also highlights the types in the visual example properly. One sec, one more commit.

Reply:
This looks good to me, thank you!

Reply:
@veox could you by any chance also apply the changes we've discussed on this PR to pygments?

Reply:
Yup, I'll find the time this week to review.

Reply:
@veox/@axic I confess I missed quite how many keywords are being generated here. Could we instead use a regular expression that uses `\d` to match digits? I realise that this will cause there to be keywords matched that aren't technically keywords in Solidity but syntax correctness is a non-goal of Rouge and if it keeps things more performant and simpler, that's a better solution from the project's perspective. What do you think?

Reply:
OK, one sec.

Reply:
Simpler is good! :) And sorry for perhaps being a little unclear: by performant I was thinking more in terms of memory than anything else. Each element is another object (or, given this is Ruby, quite a few objects) that are allocated and then garbage collected.

Thanks for the quick change!